### ISSUE TITLE
[ENH] Transfer Entropy as a pairwise time series metric

### ISSUE THREAD

**Issue by Spinachboul** [alan-turing-institute/sktime#7963]
- Currently Sktime lacks information-theoretic measures for quantifying directed relationships between time series.
- While `GrangerCausalityTests` is still under implementation , https://github.com/sktime/sktime/pull/7893 there is no implementation for Transfer Entropy, which is a model free approach that can capture both linear and non linear relationship between time series.

**Implementation Proposition**
- TE ( Transfer Entropy ) can be added as a pairwise metric int the `sktime.distances` module, following the existing design patterns of distance / similarity measures. The implementation should 
1) Create a `Transfer Entropy` class that inherits from `BasePairwiseTransformerPanel` with the following key params: 
  a. **lag_source**: int, lag order for the source time series.
  b. **lag_target**: int, lag order for the target time series.
  c. **estimator**: str, method for probab estimation (eg. "binning")
  d. **n_bins**: int, number of bins used if binning chosen
  e. **significance_test**: bool, option to perform statistical testing via surrogate data.
2) Implementing methods:
  a. `_transform(self, X, y=None)` : To return a pairwise TE matrix
  b. `_get _transfer_entropy(self, X, Y)` : for the core calculation.

**ALTERNATIVE APPROACH**
We could also interface external python packages like [PyInform](https://elife-asu.github.io/PyInform/timeseries.html#module-pyinform.transferentropy), or software packages like [IDTxl](https://github.com/pwollstadt/IDTxl) and [JIDT](https://github.com/jlizier/jidt) as backends.

**Mathematical Context**
TE can be mathematically calcuated as: 
`TE(X→Y) = H(Y₍ₜ₎|Y₍ₜ₋₁:ₜ₋ₖ₎) - H(Y₍ₜ₎|Y₍ₜ₋₁:ₜ₋ₖ₎, X₍ₜ₋₁:ₜ₋ₗ₎)`
where H  represents *Shannon Entropy*

**Comment by fkiraly** [comment]
Interesting idea!

From pyinform, there would also be mutual information and relative entropy as distances.

Regarding the parameterization, can you explain the lags and the estimator?

Lags could be added by composing with `Lag` transformer. Though I see how it would be good to have shortcuts.

Regarding the `estimator`, what is this for? Are we using it to predict and then calculate the log-loss to estimate entropy? If so, what type of estimator would it be, a tabular probabilistic regressor?

Further, `pyinform` seems "dead", it has not been maintained since 6 years. FYI @dglmoore, @jakehanson, @colemathis, if you are still active.

I would hence suggest code copy or full vendoring, with attribution of course.

Possibly clean reimplementation (but taking as inspiration) if we want to generalize to use of estimators etc.

**Comment by Spinachboul** [comment]
@fkiraly 
- Thanks for the suggestions, even I feel vendoring here makes much more sense, if the external package seems to be dead.
I think a clean implementation here along with extending to Mutual Information and KL divergence would be helpful.
- The idea behind `estimator` is to determine how we estimate probabilities for entropy calculations.
- If we use binning, it discretizes the values before computing probability distributions. But then you rightly mentioned about the use of probabilistic regressor to estimate entropy via log-loss. could we allow users to choose from both approaches, please let me know ur thoughts.

**Comment by Spinachboul** [comment]
@fkiraly Could you please assign me issue to me?
Now with your permission I will get started with the initial implementation of TE, from scratch, probably would take me while to come with a prototype due to college exams starting soon!!

**Comment by fkiraly** [comment]
@Spinachboul, great! Could you kindly paste your proposed parametrization and docstring, and explain the individual parameters? I am still not sure about what makes best sense here, so I would appreciate if we could talk this through.

**Comment by Spinachboul** [comment]
> [@Spinachboul](https://github.com/Spinachboul), great! Could you kindly paste your proposed parametrization and docstring, and explain the individual parameters? I am still not sure about what makes best sense here, so I would appreciate if we could talk this through.

Sure! I would do that, also I would come up with an implementation plan which we can disucss further
Thanks for the suggestion!!

**Comment by Spinachboul** [comment]
@fkiraly This would be the docstrings + prototype for the TE implementation
```python
class TransferEntropy(BasePairwiseTransformerPanel):
    """Calculate Transfer Entropy between pairs of time series.
    
    Transfer Entropy (TE) is an information-theoretic measure that quantifies the directed 
    transfer of information between time series. It detects both linear and non-linear 
    dependencies by measuring the reduction in uncertainty about a target time series 
    when considering the history of a source time series, beyond what can be explained 
    by the target's own history alone.
    
    Mathematically defined as:
    TE(X→Y) = H(Y(t)|Y(t-1:t-k)) - H(Y(t)|Y(t-1:t-k), X(t-1:t-l))
    
    where H represents Shannon Entropy.
    
    Parameters
    ----------
    lag_target : int, default=1
        The number of past values of the target time series to consider.
        Controls k in the formula above.
    
    lag_source : int, default=1
        The number of past values of the source time series to consider.
        Controls l in the formula above.
    
    estimator : str, default="binning"
        Method for probability estimation:
        - "binning": Equal-width binning of continuous values
        - "knn": K-nearest neighbors density estimation
        - "kernel": Kernel density estimation
    
    n_bins : int, default=10
        Number of bins used when estimator="binning".
        Higher values increase precision but require more data.
    
    k_neighbors : int, default=5
        Number of neighbors when estimator="knn".
    
    kernel_width : float, default=0.1
        Bandwidth parameter when estimator="kernel".
    
    significance_test : bool, default=False
        Whether to perform statistical significance testing.
    
    n_surrogates : int, default=100
        Number of surrogate time series to generate for significance testing.
        Only used when significance_test=True.
    
    significance_level : float, default=0.05
        P-value threshold for significance testing.
    
    normalize : bool, default=False
        If True, normalize TE values by the entropy of the target.
    
    missing_values : str, default="error"
        Strategy for handling missing values:
        - "error": Raise an error if missing values are found
        - "drop": Remove time points with missing values
        - "impute": Apply simple mean imputation
    
    Examples
    --------
    >>> from sktime.datasets import load_airline
    >>> from sktime.distances import TransferEntropy
    >>> import pandas as pd
    >>> import numpy as np
    >>> X = pd.DataFrame({
    ...     "ts1": np.random.rand(100),
    ...     "ts2": np.random.rand(100),
    ... })
    >>> te = TransferEntropy(lag_source=2, lag_target=1, estimator="binning")
    >>> te_matrix = te.fit_transform(X)
    >>> print(te_matrix)
    
    Notes
    -----
    - Higher TE values indicate stronger directed information flow from source to target.
    - TE = 0 implies conditional independence (source doesn't improve prediction of target).
    - The measure is asymmetric: TE(X→Y) ≠ TE(Y→X).
    - Requires sufficient data points for reliable estimation (rule of thumb: at least 10^(lag_source + lag_target) points).
    - Computational complexity increases with lags and estimation precision.
    
    References
    ----------
    .. [1] Schreiber, T. (2000). Measuring information transfer. Physical Review Letters, 85(2), 461.
    .. [2] Vicente, R., Wibral, M., Lindner, M., & Pipa, G. (2011). Transfer entropy—a model-free 
           measure of effective connectivity for the neurosciences. Journal of Computational Neuroscience, 30, 45-67.
    """
    
    _tags = {
        "capability:missing_values": False,
        "capability:multivariate": True,
        "capability:unequal_length": False,
    }
    
    def __init__(
        self,
        lag_target=1,
        lag_source=1,
        estimator="binning",
        n_bins=10,
        k_neighbors=5,
        kernel_width=0.1,
        significance_test=False,
        n_surrogates=100,
        significance_level=0.05,
        normalize=False,
        missing_values="error",
    ):
        self.lag_target = lag_target
        self.lag_source = lag_source
        self.estimator = estimator
        self.n_bins = n_bins
        self.k_neighbors = k_neighbors
        self.kernel_width = kernel_width
        self.significance_test = significance_test
        self.n_surrogates = n_surrogates
        self.significance_level = significance_level
        self.normalize = normalize
        self.missing_values = missing_values
        super().__init__()
        
    def _transform(self, X, y=None):
        """Calculate pairwise Transfer Entropy between all time series in X.
        
        Parameters
        ----------
        X : pd.DataFrame or nested pd.Series
            Panel data with multiple time series
        y : pd.Series, optional (default=None)
            Ignored, exists for API consistency
            
        Returns
        -------
        pd.DataFrame
            Matrix of pairwise Transfer Entropy values
        """
        # Implementation code would go here
        pass
    
    def _get_transfer_entropy(self, source, target):
        """Calculate Transfer Entropy from source to target time series.
        
        Parameters
        ----------
        source : pd.Series or np.array
            Source time series
        target : pd.Series or np.array
            Target time series
            
        Returns
        -------
        float
            Transfer Entropy value from source to target
        """
        # Core calculation code would go here
        pass
```

Please let me know your reviews on it!!

**Comment by fkiraly** [comment]
looks good - what exactly is being tested here?

**Comment by fkiraly** [comment]
nice, looking forward to a PR!
